xchange-lgo/src/main/java/info/bitrich/xchangestream/lgo/LgoLevel2BatchSubscription.java

```diff
@@ -0,0 +1,67 @@
+package info.bitrich.xchangestream.lgo;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import info.bitrich.xchangestream.lgo.domain.LgoGroupedLevel2Update;
+import info.bitrich.xchangestream.lgo.dto.LgoLevel2Update;
+import info.bitrich.xchangestream.service.netty.StreamingObjectMapperHelper;
+import io.reactivex.Observable;
+import org.knowm.xchange.currency.CurrencyPair;
+import org.knowm.xchange.dto.marketdata.OrderBook;
+import org.slf4j.*;
+
+import java.io.IOException;
+
+class LgoLevel2BatchSubscription {
+
+    private final LgoStreamingService service;
+    private final Observable<OrderBook> subscription;
+    private static final Logger LOGGER = LoggerFactory.getLogger(LgoLevel2BatchSubscription.class);
+    private CurrencyPair currencyPair;
+
+    static LgoLevel2BatchSubscription create(LgoStreamingService service, CurrencyPair currencyPair) {
+        return new LgoLevel2BatchSubscription(service, currencyPair);
+    }
+
+    private LgoLevel2BatchSubscription(LgoStreamingService service, CurrencyPair currencyPair) {
+        this.service = service;
+        this.currencyPair = currencyPair;
+        subscription = createSubscription();
+    }
+
+    Observable<OrderBook> getSubscription() {
+        return subscription;
+    }
+
+    private Observable<OrderBook> createSubscription() {
+        final ObjectMapper mapper = StreamingObjectMapperHelper.getObjectMapper();
+        return service
+                .subscribeChannel(LgoAdapter.channelName("level2", currencyPair))
+                .map(s -> mapper.readValue(s.toString(), LgoLevel2Update.class))
+                .scan(new LgoGroupedLevel2Update(), (acc, s) -> {
+                    if (s.getType().equals("snapshot")) {
+                        acc.applySnapshot(s.getBatchId(), s.getData());
+                        return acc;
+                    }
+                    if (acc.getLastBatchId() + 1 != s.getBatchId()) {
+                        LOGGER.warn("Wrong batch id. Expected {} got {}.", acc.getLastBatchId() + 1, s.getBatchId());
+                        resubscribe();
+                    }
+                    acc.applyUpdate(s.getBatchId(), s.getData());
```

Comment:
Correct. I couldn't decide what I wanted to do here. 
On the other hand, what is the best way to avoid emitting this one, without disposing the observable ? 

Reply:
How about wrapping the orderbook on the stream with a simple object containing the orderbook and a `dirty` flag. This flag gets set to `true` in the `scan` if you get an out of sequence update, and can therefore be `filter`ed immediately afterwards and then `map`ped to the order book.